src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs

```diff
@@ -195,9 +195,8 @@ private IServiceCollection BuildHostingServices()
             services.AddTransient<IHttpContextFactory, HttpContextFactory>();
             services.AddOptions();
 
-            var diagnosticSource = new DiagnosticListener("Microsoft.AspNetCore");
-            services.AddSingleton<DiagnosticSource>(diagnosticSource);
-            services.AddSingleton<DiagnosticListener>(diagnosticSource);
+            services.AddSingleton<DiagnosticListener>(provider => new DiagnosticListener("Microsoft.AspNetCore"));
```

Comment:
Also we don't mutate `DiagnosticSource` so it might be fine to return different instances.